File: minimum-depth-of-binary-tree/mini_bt.go
package minBt

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type qn struct {
	node  *TreeNode
	depth int
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	queue := make([]qn, 0, 1024)

	queue = append(queue, qn{root, 1})

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		if node.node.Left == nil && node.node.Right == nil {
			return node.depth
		}

		if node.node.Left != nil {
			queue = append(queue, qn{node.node.Left, node.depth + 1})
		}

		if node.node.Right != nil {
			queue = append(queue, qn{node.node.Right, node.depth + 1})
		}
	}

	return 0
}

File: minimum-depth-of-binary-tree/python/mini_bt.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        if root is None:
            return 0
        count = 0
        
        return self._depth(root, count)
        
    def _depth(self, node: TreeNode, count: int) -> int:
        if node is None:
            return pow(10, 5) + 1
        
        count += 1
        
        
        if node.left is None and node.right is None:
            return count
        
        
        count = min(
            self._depth(node.left, count),
            self._depth(node.right, count),
        )
        
        return count
